mlir/lib/Dialect/Linalg/Transforms/DecomposeGenericByUnfoldingPermutation.cpp

```diff
@@ -159,6 +160,16 @@ LogicalResult DecomposeProjectedPermutation::matchAndRewrite(
     auto map = op.getMatchingIndexingMap(&opOperand);
     if (!map.isProjectedPermutation(false))
       return failure();
+
+    // If we have any inputs that aren't ranked tensor types, reject
+    // the pattern.
+    if (!dyn_cast<RankedTensorType>(opOperand.get().getType()))
+      return rewriter.notifyMatchFailure(
+          opOperand.get().getLoc(),
+          llvm::formatv("Expected operand #{0} to be "
+                        "ranked tensor of any type values, but got {1}",
+                        opOperand.getOperandNumber(),
+                        opOperand.get().getType()));
```

Comment:
[nit] `notifyMatchFailure` messages tend to be fairly simple. My preference would be to keep this one simple as well and avoid this formatting.